Improve texture modifier documentation

Wuzzy2 opened the pull request:
This PR updates the texture modifier documentation by making it somewhat more accurate and precise.

Work based on <https://docs.luanti.org/for-creators/api/texture-modifiers/>, the texts are basically copied from there with minor changes. Thanks to @appgurueu.

(Ideally, it would have been better if those changes would have gone to `lua_api.md` first, but … oh well)

## To do

* [DONE] Wait for #16735 to get merged
* [DONE] Update texture mofidier documentation to correctly document integer ranges, following the spirit of #16735

## How to test

Read the changes. ;-)

j-r commented on doc/lua_api.md:
```diff
@@ -630,37 +646,46 @@ Example:
+texture to the maximum value (255).
+
+As the red, green and blue channels aren't premultiplied with alpha in PNGs,
+this might reveal hidden colors of otherwise transparent portions of an image.
```
I think this affects other operations like `[invert`, too. Perhaps make this a general comment with a suitable link to e.g. https://en.wikipedia.org/wiki/Alpha_compositing (not sure if that is a good resource, though).

j-r commented on doc/lua_api.md:
```diff
@@ -588,35 +588,51 @@ on top of `cobble.png`.
+* `<w>`: width of the resulting texture
+* `<h>`: height of the resulting texture
+* `<textures>: Colon (`:`)-separated list of locations `x`, `y` (both positive
+  integers) and escaped texture modifiers texture to blit in the form
```
`escaped texture modifiers texture` isn't clear. Perhaps add something like `<texture> can contain texture modifiers, but these must be escaped according to the rules in the Escaping section above` below the next line.

Or introduce the term in the introduction?

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -588,35 +588,51 @@ on top of `cobble.png`.
+* Use the third crack progression (`2` because of 0-indexing) ...
+* ... and draw it on a stone texture consisting of `1` tile
+
+    default_lava_source_animated.png^[crack:1:8:2
```
No, this is correct. But the leading `1` can be omitted.

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -670,44 +695,62 @@ Example:
 
-`<t>` can be a number (between 0 and 7) or a transform name.
-Rotations are counter-clockwise.
+`<t>` is the concatenation of either numbers or names identifying
```
Done. But all the number/integer confusion is bound to be fixed in #16735 before.

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -588,35 +588,51 @@ on top of `cobble.png`.
 
-Example:
+Note: This always scales the crack to the size of the base texture (or the tiles
+of the base texture, if tilesize is provided).
```
I've completely rewritten the crack section.

Wuzzy2 commented:
Addressed your comments.
I also completely rewrote the `[crack` documentation because it turns out that `[crack` behaves much differently than we previously believed (determined by testing).

So my recommended use of `[crack` is the 2-parameter version both for normal and animated textures. The 3-parameter version seems rather esoteric because it creates a grid of cracks. I suppose this could be used for tilesheets.

Wuzzy2 committed:
Improve texture modifier documentation

Wuzzy2 commented:
Since https://github.com/luanti-org/luanti/pull/16735 got merged, I have updated this PR to incorporate the latest changes. This PR is no longer a draft and ready.

Please note the documentation has changed significantly from the original version due to the changes by #16735.

cx384 commented on doc/lua_api.md:
```diff
@@ -706,42 +706,72 @@ on top of `cobble.png`.
+* `grid`: Creates a square grid of multiple cracks, with
+  side length `grid`, e.g. the number 3 creates a 3x3 grid of cracks
+  (integer [imagesize], default 1)
+* `vertical_frames`: Amount of times the grid created by `tile_count`
```
What does ""created by `tile_count`" mean here?
I can't find the term `tile_count` anywere.

Wuzzy2 replied:
It should have been `grid`.

cx384 commented on doc/lua_api.md:
```diff
@@ -706,42 +706,72 @@ on top of `cobble.png`.
 
-    default_cobble.png^[crack:10:1
+* `vertical_frames`: Amount of times the crack will be blit on the
+  base texture, stacked vertically (integer [imagesize])
```
This sounds confusing to me. Can't you just write something like
Amount of times the crack frame will be applied repeatedly in vertical direction.

cx384 commented on doc/lua_api.md:
```diff
@@ -706,42 +706,72 @@ on top of `cobble.png`.
-  (integer [imagesize], often `1`)
-* `<p>`: crack animation frame
-  (integer [imageframe], counting starts at 0)
+More precisely, it overlays a scaled frame of the crack texture,
```
```suggestion
It overlays a scaled frame of the crack texture,
```

cx384 commented on doc/lua_api.md:
```diff
@@ -706,42 +706,72 @@ on top of `cobble.png`.
-is to the left and/or upwards of the origin of the target texture.
+* Cracked lava
+* Use the 3rd crack progression ...
+* ... and draw it 8 times on top of the base texture
```
```suggestion
* Use the 3rd crack progression and draw it 8 times on top of the base texture
```

cx384 commented on doc/lua_api.md:
```diff
@@ -706,42 +706,72 @@ on top of `cobble.png`.
+
+* Cracked stone
+* Use the third crack progression (`2` because of 0-indexing) ...
+* ... and draw it once on a stone texture
```
```suggestion
* Use the third crack progression (`2` because of 0-indexing)
  and draw it once on a stone texture
```

cx384 commented on doc/lua_api.md:
```diff
@@ -706,42 +706,72 @@ on top of `cobble.png`.
+
+* `<w>`: width of resulting texture (integer [imagesize])
+* `<h>`: height of resulting texture (integer [imagesize])
+* `<textures>: Colon-(`:`)-separated list of locations `x`, `y` and textures to
```
```suggestion
* `<textures>`: Colon-separated (`:`) list of locations `x`, `y` and textures to
```

cx384 commented on doc/lua_api.md:
```diff
@@ -706,42 +706,72 @@ on top of `cobble.png`.
+
+`x` and `y` are integers [imageframe].
+A `<texture>` (in `<textures>`) can contain texture modifiers, but these must
+be escaped according to the rules in the Escaping section above.
```
```suggestion
be escaped according to the rules in [Escaping](#Escaping).
```

cx384 commented on doc/lua_api.md:
```diff
@@ -758,37 +788,46 @@ Example:
 
-Makes the base image transparent according to the given ratio.
+This multiplies the alpha value of each pixel of the base texture
+with `ratio/255` and rounds properly afterwards.
```
what does "properly" mean? up? down?

Wuzzy2 replied:
It's the "round half up rule". Fixed.

cx384 commented on doc/lua_api.md:
```diff
@@ -758,37 +788,46 @@ Example:
+* `mode` is a string which may contain the characters `r`, `g`, `b` and `a`.
+
+The channels corresponding to the occurring characters (red, green, blue
+and alpha) will be inverted (set to `255-value`).
```
```suggestion
and alpha) will be inverted (set to `255 - value`).
```

cx384 commented on doc/lua_api.md:
```diff
@@ -758,37 +788,46 @@ Example:
 
-Makes the texture completely opaque.
+Makes the texture completely opaque by setting the alpha channel of the base
+texture to the maximum value (255).
```
```suggestion
texture to the maximum value (`255`).
```

cx384 commented on doc/lua_api.md:
```diff
@@ -758,37 +788,46 @@ Example:
+Makes the texture completely opaque by setting the alpha channel of the base
+texture to the maximum value (255).
+
+As the red, green and blue channels aren't premultiplied with alpha in PNGs,
```
```suggestion
As the red, green and blue channels aren't pre-multiplied with alpha in PNGs,
```

cx384 commented on doc/lua_api.md:
```diff
@@ -798,6 +837,13 @@ Example:
 Convert the given color to transparency.
 
+Pixels of the base texture having the exact same RGB color will have their
+alpha value set to 0.
```
```suggestion
alpha value set to `0`.
```

cx384 commented on doc/lua_api.md:
```diff
@@ -808,38 +854,49 @@ Example:
+
+As a formula:
+
+    9 * max(4, min(64, 2^ceil(log_2(max(d)))))
```
```suggestion
    `9 * max(4, min(64, 2^ceil(log_2(max(d)))))`
```

Wuzzy2 replied:
No. In Markdown, indenting text with 4 spaces makes it a code block. https://daringfireball.net/projects/markdown/syntax#precode

cx384 commented on doc/lua_api.md:
```diff
@@ -808,38 +854,49 @@ Example:
     [inventorycube{grass.png{dirt.png&grass_side.png{dirt.png&grass_side.png
 
-Creates an inventorycube with `grass.png`, `dirt.png^grass_side.png` and
+Creates a cube with `grass.png`, `dirt.png^grass_side.png` and
```
It does not actually create a cube so inventorycube is more correct here.

Wuzzy2 replied:
I’ll go with “Renders a cube” cuz “inventorycube” is a strange word.

cx384 commented on doc/lua_api.md:
```diff
@@ -886,35 +943,48 @@ Example:
+
+* `texture` is an escaped texture modifier
+
+Applies a *bitwise and* (`bit.band`) to all RGBA values of `texture` and
```
```suggestion
Applies a *bitwise and* to all RGBA values of `texture` and
```
It doesn't use `bit.band`.

cx384 commented on doc/lua_api.md:
```diff
@@ -886,35 +943,48 @@ Example:
+* Defaults to the alpha of `color` if omitted;
+* If it's an integer from 0 (only pixel color) to 255 (only color), it is
+  directly used as interpolation ratio: The resulting color of a pixel is
+  ratio times color plus `(255-ratio)` times pixel color;
```
```suggestion
  `ratio` times `color` plus `255 - ratio` times pixel color
```

cx384 commented on doc/lua_api.md:
```diff
@@ -886,35 +943,48 @@ Example:
+* If it's an integer from 0 (only pixel color) to 255 (only color), it is
+  directly used as interpolation ratio: The resulting color of a pixel is
+  ratio times color plus `(255-ratio)` times pixel color;
+* If it's the string "alpha", the texture pixel's alpha value determines the
```
```suggestion
* If it's the string `"alpha"`, the texture pixel's alpha value determines the
```

cx384 commented on doc/lua_api.md:
```diff
@@ -886,35 +943,48 @@ Example:
+Interpolates between color and the pixel colors of the base texture as specified
+by the `ratio`:
+
+* Defaults to the alpha of `color` if omitted;
```
```suggestion
* Defaults to the alpha of `color` if omitted
```

cx384 commented on doc/lua_api.md:
```diff
@@ -936,7 +1006,8 @@ and -120° is blue.
 
-Multiplies texture colors with the given color.
+Multiplies the RGB values of the base texture per pixel with the RGB
+values of color; the `alpha` value of `color` is ignored.
```
```suggestion
values of `color`; the `alpha` value of `color` is ignored.
```

cx384 commented on doc/lua_api.md:
```diff
@@ -886,35 +943,48 @@ Example:
+by the `ratio`:
+
+* Defaults to the alpha of `color` if omitted;
+* If it's an integer from 0 (only pixel color) to 255 (only color), it is
```
```suggestion
* If it's an integer from 0 (only pixel color) to 255 (only `color`), it is
```

Wuzzy2 replied:
No, here "color" is meant as the English word.

cx384 replied:
I thought this would make it clearer, since "color" here refers to the specified `color`, right?
Otherwise, the distinction between " only pixel color" and "only color" may be harder to understand.

Wuzzy2 committed:
Fix some formatting issues in lua_api.md

Wuzzy2 committed:
Docs: Clarify rounding rule of [opacity

cx384 commented on doc/lua_api.md:
```diff
@@ -886,35 +943,47 @@ Example:
+
+Colorizes the textures with the given color.
+
+Interpolates between color and the pixel colors of the base texture as specified
```
```suggestion
Interpolates between `color` and the pixel colors of the base texture as specified
```
Here too?

cx384 commented on doc/lua_api.md:
```diff
@@ -758,37 +787,47 @@ Example:
-Makes the base image transparent according to the given ratio.
+This multiplies the alpha value of each pixel of the base texture
+with `ratio/255` and rounds the result afterwards (using the
+"round half up" rule).
```
```suggestion
with `ratio/255` and rounds the result to the closest integer afterwards.
```

cx384 reviewed (commented):
Look mostly good now, I think.
To me, it seems a bit too elaborated at some places, but it adds important information that was missing before.

For some things I'm not sure if we want it to be documented, like the `[opacity` rounding thing, or the `[inventorycube` formula.

Though, it's at least better than before.

Wuzzy2 committed:
Minor clarifications in texture modifier docs

Wuzzy2 commented:
Done.

SmallJoker commented on doc/lua_api.md:
```diff
@@ -706,42 +706,71 @@ on top of `cobble.png`.
-`crack` draws it normally, while `cracko` lays it over, keeping transparent
-pixels intact.
+With `crack`, the crack will be overlaid over the whole base
+texture, while with `cracko`, the crack will only be overlaid
```
```suggestion
With `[crack`, the crack will be overlaid over the whole base
texture, while with `[cracko`, the crack will only be overlaid
```

.. perhaps, to be more precise?

SmallJoker commented on doc/lua_api.md:
```diff
@@ -706,42 +706,71 @@ on top of `cobble.png`.
+over fully opaque base texture regions.
 
-Example:
+Parameters if `grid` is omitted:
```
Does this need special treatment? Just specify that `grid` defaults to `1` if not specified. The subsequent explanations would then still hold true. Avoids duplication.

SmallJoker commented on doc/lua_api.md:
```diff
@@ -808,38 +853,49 @@ Example:
+
+The resulting image will be 9 times the nearest power of 2 large, enough to
+contain the dimensions of the largest image, clamped to a range of at least
+4 and at most 64.
```
Do we really want to document this? This seems oddly specific and I'd rather not want modders to rely on this.

j-r replied:
If one wants to overlay something on the resulting image or combine it with other things it is unfortunately necessary to know the exact size of the result, because currently most modifiers take pixel coordinates.

It would be really great to have a size agnostic modifier language, because different clients may be using different texture packs with different resolutions. Something like accepting coordinates with a decimal point as relative, e.g. `[resize:1.x2.]` doubling height.

Wuzzy2 replied:
Yeah, I have to agree with @j-r here, the size needs to be documened.

If you want to reserve the right to arbitrarily change the output size of this element in future: Please don’t, you will just break mods that assume the current size (regardless of whether it is documentated or not).

If you *want* to change the output size in future, you can just add more optional parameters to this element (with the current hardcoded size as default), so backwards-compatibility is preserved.

appgurueu replied:
Perhaps we should add an option for specifying a target resolution of `[inventorycube`.

I don't think there's any harm in keeping the current formula though. Maybe add a comment to the code that it should not be changed.

"Layouting" capabilities / resolution-agnostic texture modifiers more generally are a trickier discussion for another issue (related: my opinions on texture modifier design #11588), and also ties into SSCSM: The most powerful solution is obviously if you can do whatever image manipulation you desire (#16794).

SmallJoker replied:
The current formula (= the returned image) depends on the node texture size, hence I doubt this information is of any use to modders because the resolution cannot be retrieved in any helpful way. `[resize` will always be necessary to cover all cases.

appgurueu replied:
I don't think it's terribly important to document this. Most (all?) modders indeed won't get any use out of it. At the same time, I think there's no point in ever changing the current formula as it is. Indeed I think it's likely that people have come to rely on it, if only purely by experimentation (which probably means their code implicitly relies on, say, 16x16 node textures).

What might be important to mention still is that `[inventorycube` resolution grows with the node texture resolution, ensuring that `[inventorycube` is crisp (and implying that modders expecting a fixed resolution should `[resize`).

Wuzzy2 replied:
Simplified the description.

j-r replied:
> The current formula (= the returned image) depends on the node texture size, hence I doubt this information is of any use to modders because the resolution cannot be retrieved in any helpful way. `[resize` will always be necessary to cover all cases.

Yes, but if the input texture has already been resized to a known size to apply some size dependent modifier then it should not be necessary to resize the output of `[inventorycube` again. (Though I haven't tested whether complex input textures actually work with the special escaping here, but it should, I think.)

SmallJoker commented on doc/lua_api.md:
```diff
@@ -886,35 +942,47 @@ Example:
+* Defaults to the alpha of `color` if omitted
+* If it's an integer from 0 (only base texture color) to 255 (only `color`),
+  it's directly used as interpolation ratio: The resulting color of a pixel is
+  `ratio` times `color` plus `(255 - ratio)` times base texture color;
```
```suggestion
* If it's an integer from 0 (only base texture color) to 255 (only `color`),
  the resulting color of a pixel is `ratio * color + (255 - ratio) * base_tex_color`.
```

SmallJoker commented on doc/lua_api.md:
```diff
@@ -886,35 +942,47 @@ Example:
+  it's directly used as interpolation ratio: The resulting color of a pixel is
+  `ratio` times `color` plus `(255 - ratio)` times base texture color;
+* If it's the string `"alpha"`, the texture pixel's alpha value determines the
+  ratio per pixel
```
I'd prefer to clearly mention `ratio` here because "alpha" is most likely associated with "color" and not "ratio".
```suggestion
* If `ratio = "alpha"`, the texture pixel's alpha value determines the
  ratio per pixel
```

SmallJoker commented on doc/lua_api.md:
```diff
@@ -706,42 +706,71 @@ on top of `cobble.png`.
+
+* Blits a 3x3 grid of crack textures, each at frame 0, on `example.png`
+
+#### `[combine:<w>x<h>:<textures>`
```
I propose to name `textures`  `parts`, `sections` or `insertions` because `textures` is not just concat' texture names

appgurueu replied:
`parts` sounds good to me. maybe `blits`?

SmallJoker reviewed (commented):
I hope this does not contradict with earlier reviews.

appgurueu commented on doc/lua_api.md:
```diff
@@ -866,9 +922,9 @@ Examples:
     texture.png^[fill:8x8:4,4:red
 
-#### `[lowpart:<percent>:<file>`
+#### `[lowpart:<percent>:<texture>`
```
shouldn't there be a "\* `texture` is an escaped texture modifier" here?

appgurueu commented on doc/lua_api.md:
```diff
@@ -886,35 +942,47 @@ Example:
+If a pixel of the base texture is out of bounds on texture, it is preserved.
 
-The mask is applied using binary AND.
+Masking is associative and commutative if all involved textures have the same
```
not sure why i found this important to mention at the time, i think this can be omitted. it's clear from the definition.

appgurueu commented on doc/lua_api.md:
```diff
@@ -886,35 +942,47 @@ Example:
-the word "`alpha`", then each texture pixel will contain the RGB of
-`<color>` and the alpha of `<color>` multiplied by the alpha of the
-texture pixel.
+
```
```suggestion
```

Wuzzy2 committed:
Clean up texture modifier documentation

appgurueu commented on doc/lua_api.md:
```diff
@@ -886,35 +942,47 @@ Example:
+
+Colorizes the textures with the given color.
+
+Interpolates between `color` and the pixel colors of the base texture as specified
```
I think we need to be careful here. If alpha were interpolated, we would expect `[colorize` to make images semitransparent normally, which we don't want. What exactly happens with `alpha`?

Peeking at the implementation (`apply_colorize` in `imagesource.cpp`): Something funny.

Let $r$ be the "effective ratio" (either the given number, or the alpha channel of the color if none is given).

In all cases, only pixels with alpha > 0 are touched.

If we have a ratio that is not alpha, or if the ratio is alpha but the alpha of the color is less than 255, we get an interpolation that *includes alpha*. I don't think this makes much sense for actual semitransparent pixels but that's how it is. It follows that the only real sensible usage we should recommend in this case is with a base image that does not use semitransparency, and a color that has an alpha of 255.

If the ratio is alpha, *and* the color has an alpha of 255, the alpha channel of the original image is kept, but the color replaced.

This seems quite odd to me. Maybe I'm misreading something, maybe there's a bug, maybe the documentation needs to be fixed. In any case I think something funky is going on, and I wouldn't bet on myself from some years back having analyzed this properly.

Wuzzy2 replied:
I would say this behavior is an outright bug. I have tested this in a test mod and can confirm what you said.

The point of `colorize` is to change the color of a base image. That’s it. That was always it’s “selling point”. Therefore, it should not randomly modify alpha IMHO. I think a functional `colorize` implementation should guarantee that the alpha values of the output image are always the same as from the input image. Especially the fact that the "funky" behavior is only triggered for very specific values is pretty sus.

Exception: When the `"alpha"` ratio is used. I noticed that `"alpha"` makes no sense when you also keep the alpha channel in the output image. The lower the alpha, the lesser visible the lack of color change becomes. Not very useful. What I think was intended here is that this parameter means, that the alpha channel of the base image is only used as a separate source of information, not also as actual base image input. The "true" base image would thus only be the raw color or it, assuming alpha of 255 everywhere, so the output image is always fully opaque.

In any case, for the documentation I’ve decided to add several warnings and guide the modder to only use specific values for now. In the long runs, `[colorize` needs to be fixed IMHO.

Wuzzy2 committed:
Docs update: [inventorycube, [lowpart, [colorize

SmallJoker reviewed (approved):
Looks good. Thank you for the updates.

appgurueu commented on doc/lua_api.md:
```diff
@@ -962,20 +952,22 @@ which is assumed to be a tilesheet with dimensions
 * If `ratio = "alpha"`, the texture pixel's alpha value determines the
-  ratio per pixel
+  ratio per pixel (NOT RECOMMENDED)
+* NOTE: Due to bugs, we currently can only recommend the base image to be fully
```
👍 on the "not recommended's" *but* I propose this be relaxed to "we currently can only recommend the base image to **avoid any semitransparent pixels** (alpha > 0, < 255)".

alpha = 0 is fine and will continue to work as before when we fix this

appgurueu replied:
https://github.com/luanti-org/luanti/pull/16753/changes/b1c1f770adf9861e4fb9da0cfc693f4d9113399b

appgurueu committed:
Relax [colorize base image transparency constraint
